### PR TITLE
Features/340 gaussian nb

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

Initiating submodule naive_bayes and class GaussianNB().

Implementation of Gaussian Naive Bayes classification along the scikit-learn lines. Known issues listed below.

Issue/s resolved: #340 

## Changes proposed:
- New submodule naive_bayes (on the same level as cluster and regression)
- New classifier  naive_bayes.GaussianNB()
- Implemented functions GaussianNB().fit(), GaussianNB().predict(), GaussianNB().predict_proba() plus helper functions
- Implemented test_gaussiannb() for local and distributed mode. In distributed mode, only the case where split=0 is tested for the time being (labels are 1D). Different split configurations require resplit to work from 1 to 0 (cf. #425).

## Known issues:
- #351 , #450, #457, #458, #468, #490
- Address case where X.split != 0. Redistribute?  (needs #425)
- ~~Adapt documentation, conventions from sklearn to HeAT (I know about those missing dunderscores!)~~ DONE
- Create demo

## Type of change

- New feature (non-breaking change which adds functionality)
- Documentation update

## Due Diligence

- [ ] All split configurations tested- NOT YET: Different split configurations require resplit to work from 1 to 0 (cf. #425).
- [ ] Multiple dtypes tested in relavent functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/naive_bayes/gaussianNB.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+    >>> print(clf_pf.predict([[-0.8, -1]]))
+    [1]
+    """
+
+    def __init__(self, priors=None, var_smoothing=1e-9):
+        self.priors = priors
+        self.var_smoothing = var_smoothing
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit Gaussian Naive Bayes according to X, y
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: Thanks, done.

File: heat/core/naive_bayes/gaussianNB.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+        # (weighted) number of observations. This is achieved by combining
+        # the sum-of-squared-differences (ssd)
+        old_ssd = n_past * var
+        new_ssd = n_new * new_var
+        total_ssd = old_ssd + new_ssd + (n_new * n_past / n_total) * (mu - new_mu) ** 2
+        total_var = total_ssd / n_total
+
+        return total_mu, total_var
+
+    def partial_fit(self, X, y, classes=None, sample_weight=None):
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: The internal __partial_fit function has an extra kwarg _refit that one can play around with (if True, throw previous fits away and start over). I'm going to keep it for now but you're right that this is an area that can be simplified.

File: heat/core/naive_bayes/gaussianNB.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+        if sample_weight is not None:
+            if not isinstance(sample_weight, ht.DNDarray):
+                raise ValueError(
+                    "sample_weight needs to be a ht.DNDarray, but was {}".format(
+                        type(sample_weight)
+                    )
+                )
+        classes = ht.unique(y, sorted=True)
+        if classes.split is not None:
+            gathered_classes = ht.resplit(classes, axis=None)
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: I'm sure I've done this already. Must have gotten lost in a stash. Thanks for noticing.

File: heat/core/naive_bayes/gaussianNB.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+        y : array-like, shape (n_samples,)
+            Target values.
+        sample_weight : array-like, shape (n_samples,), optional (default=None)
+            Weights applied to individual samples (1. for unweighted).
+            .. versionadded:: 0.17
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: LOL, good catch, one of those sklearn references.

File: heat/core/naive_bayes/gaussianNB.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        classes = ht.unique(y, sorted=True)
+        if classes.split is not None:
+            gathered_classes = ht.resplit(classes, axis=None)
+            return self.__partial_fit(
+                X, y, gathered_classes, _refit=True, sample_weight=sample_weight
+            )
+        return self.__partial_fit(X, y, classes, _refit=True, sample_weight=sample_weight)
+
+    # @staticmethod
+    def __check_partial_fit_first_call(clf, classes=None):
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: Right, thanks.

File: heat/core/naive_bayes/gaussianNB.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+        Perform classification on an array of test vectors X.
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+        Returns
+        -------
+        C : ndarray of shape (n_samples,)
+            Predicted target values for X
+        """
+        # check_is_fitted(self) #TODO sanitation module
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/naive_bayes/gaussianNB.py]: cleaned up

**Comment by ClaudiaComito** [comment]
> I mentioned a few things that need some work. I also made a list of things which still need to be done:
> 
>     * isfinite
> 
>     * broadcast_arrays
> 
>     * sign
> 
>     * atleast_2d
> 
> 
> Are these things impediments on the usage of this for basic cases? or are these things which are typically only used by more advanced users?

Thanks @coquelin77, most of these are part of the general sanitation/validation effort (except np.sign). Basic cases run fine without them, but they're on my list. I'll file Issues now.

File: heat/naive_bayes/gaussianNB.py [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+
+class GaussianNB:
+    """
+    Gaussian Naive Bayes (GaussianNB)
+    Can perform online updates to model parameters via :meth:`partial_fit`.
+    For details on algorithm used to update feature means and variance online,
+    see Stanford CS tech report STAN-CS-79-773 by Chan, Golub, and LeVeque:
+        http://i.stanford.edu/pub/cstr/reports/cs/tr/79/773/CS-TR-79-773.pdf
+    Read more in the :ref:`User Guide <gaussian_naive_bayes>`.
+    Parameters
```
Comment (coquelin77) [inline heat/naive_bayes/gaussianNB.py]: line breaks before parameters, attributes, examples

File: heat/naive_bayes/gaussianNB.py [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+    def __update_mean_variance(n_past, mu, var, X, sample_weight=None):
+        """
+        Compute online update of Gaussian mean and variance.
+        Given starting sample count, mean, and variance, a new set of
+        points X, and optionally sample weights, return the updated mean and
+        variance. (NB - each dimension (column) in X is treated as independent
+        -- you get variance, not covariance).
+        Can take scalar mean and variance, or vector mean and variance to
+        simultaneously update a number of independent Gaussians.
+        See Stanford CS tech report STAN-CS-79-773 by Chan, Golub, and LeVeque:
```
Comment (coquelin77) [inline heat/naive_bayes/gaussianNB.py]: Move to a references section. Use IEEE standard for the citation as well please. Also if this is the same paper as referenced in the main docs for GNB then you can only cite it there and point to it from here.

File: heat/naive_bayes/gaussianNB.py [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+        ``partial_fit`` on ``clf``. In that case the ``classes_`` attribute is also
+        set on ``clf``.
+        """
+        if getattr(self, "classes_", None) is None and classes is None:
+            raise ValueError("classes must be passed on the first call " "to partial_fit.")
+
+        elif classes is not None:
+            unique_labels = classes
+            if getattr(self, "classes_", None) is not None:
+                if not ht.equal(self.classes_, unique_labels):
```
Comment (coquelin77) [inline heat/naive_bayes/gaussianNB.py]: could these two be combined? or maybe it could be something like this
```python
if getattr(self, "classes_", None) is None:
    self.classes_ = unique_labels
    return True
if not ht.equal(self.classes_, unique_labels):
    raise ValueError("...")
return False
```

File: heat/naive_bayes/gaussianNB.py [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+        -------
+        res : ndarray
+            The result, ``np.log(np.sum(np.exp(a)))`` calculated in a numerically
+            more stable way. If `b` is given then ``np.log(np.sum(b*np.exp(a)))``
+            is returned.
+        sgn : ndarray
+            If return_sign is True, this will be an array of floating-point
+            numbers matching res and +1, 0, or -1 depending on the sign
+            of the result. If False, only one result is returned.
+        See Also
```
Comment (coquelin77) [inline heat/naive_bayes/gaussianNB.py]: line breaks, as well as the TODO within these docs

File: heat/core/statistics.py [hunk 11]
```diff
@@ -319,34 +319,37 @@ def average(x, axis=None, weights=None, returned=False):
...
+            wgt = factories.empty(wgt_lshape, dtype=weights.dtype, device=x.device)
+            wgt._DNDarray__array[wgt_slice] = weights._DNDarray__array
+            wgt = factories.array(wgt._DNDarray__array, is_split=wgt_split)
+        else:
+            if x.comm.is_distributed():
+                if x.split is not None and weights.split != x.split and weights.numdims != 1:
+                    raise NotImplementedError(
+                        "weights.split does not match data.split: not implemented yet."
+                    )
+                # fix after Issue #425 is solved
```
Comment (coquelin77) [inline heat/core/statistics.py]: too many tabs and 1 line below this it looks like there is dead code

File: heat/core/statistics.py [hunk 12]
```diff
@@ -1227,7 +1230,7 @@ def reduce_vars_elementwise(output_shape_i):
 
         var_shape = list(var.shape) if list(var.shape) else [1]
 
-        var_tot = factories.zeros(([x.comm.size, 2] + var_shape), device=x.device)
+        var_tot = factories.zeros(([x.comm.size, 2] + var_shape), dtype=x.dtype, device=x.device)
         n_tot = factories.zeros(x.comm.size, device=x.device)
```
Comment (coquelin77) [inline heat/core/statistics.py]: possible dtype problem here.

File: heat/naive_bayes/gaussianNB.py (hunk 9 above)
Comment (coquelin77) [reply heat/naive_bayes/gaussianNB.py]: thoughts?

File: heat/naive_bayes/gaussianNB.py [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+        total_mu : ht.tensor of shape (number of Gaussians,)
+            Updated mean for each Gaussian over the combined set.
+        total_var : ht.tensor of shape (number of Gaussians,)
+            Updated variance for each Gaussian over the combined set.
+        """
+        if X.shape[0] == 0:
+            return mu, var
+
+        # Compute (potentially weighted) mean and variance of new datapoints
+        # TODO:Issue #351 allow weighted average across multiple axes
```
Comment (coquelin77) [inline heat/naive_bayes/gaussianNB.py]: isnt this done now?

File: heat/naive_bayes/gaussianNB.py (hunk 9 above)
Comment (ClaudiaComito) [reply heat/naive_bayes/gaussianNB.py]: Sorry I missed  your first comment, you're right, I've updated according to your suggestion.

File: heat/naive_bayes/gaussianNB.py (hunk 8 above)
Comment (ClaudiaComito) [reply heat/naive_bayes/gaussianNB.py]: Added references both in main gnb documentation and in __update_mean_variance as I suspect the latter will be needed in other classifiers as well and move out to independent module eventually.

File: heat/core/statistics.py (hunk 11 above)
Comment (ClaudiaComito) [reply heat/core/statistics.py]: too many tabs (black's doing) or too many nested conditional statements? 

if former:
   :woman_shrugging:
else:
   I'll think about it

File: heat/core/statistics.py (hunk 12 above)
Comment (ClaudiaComito) [reply heat/core/statistics.py]: is var() supposed to return a float32? Even if x is float64?

Comment (coquelin77) [reply heat/core/statistics.py]: it should not. i guess i missed a couple dtype calls here. can you add them for me? it should be just in this spot

File: heat/core/statistics.py (hunk 11 above)
Comment (coquelin77) [reply heat/core/statistics.py]: i was more confused about the structure here. but after sleeping on it i see what you meant. i read it now with the meaning of after issue #425 is solved then the if statement above these lines and the raise within will both be removed. i misunderstood it before i think

File: heat/core/statistics.py (hunk 12 above)
Comment (ClaudiaComito) [reply heat/core/statistics.py]: That's what I thought, so is this the possible dtype problem you were talking about? I guess I just misunderstood your first comment

Comment (coquelin77) [reply heat/core/statistics.py]: i just also realized that the dtype of n_tot doesnt matter because it is only used internally

**Comment by coquelin77** [comment]
good work!
